src/NumericalAlgorithms/LinearOperators/PartialDerivatives.tpp

```diff
@@ -7,25 +7,212 @@
+  for (size_t deriv_index = 0; deriv_index < Dim; ++deriv_index) {
+    for (size_t d = 0; d < Dim; ++d) {
+      gsl::at(gsl::at(indices, d), deriv_index) =
+          InverseJacobian<DataVector, Dim, Frame::Logical,
+                          DerivativeFrame>::get_storage_index(d, deriv_index);
+    }
+  }
+
+  for (size_t component_index = 0;
+       component_index < number_of_independent_components; ++component_index) {
+    for (size_t deriv_index = 0; deriv_index < Dim; ++deriv_index) {
+      lhs.set_data_ref(pdu, num_grid_points);
+      // clang-tidy: const cast is fine since we won't modify the data and we
+      // need it to easily hook into the expression templates.
+      logical_du.set_data_ref(
+          const_cast<double*>(  // NOLINT
+              gsl::at(logical_partial_derivatives_of_u, 0)) +  // NOLINT
+              component_index * num_grid_points,
+          num_grid_points);
+      lhs = (*(inverse_jacobian.begin() + gsl::at(indices[0], deriv_index))) *
+            logical_du;
+      for (size_t logical_deriv_index = 1; logical_deriv_index < Dim;
+           ++logical_deriv_index) {
+        // clang-tidy: const cast is fine since we won't modify the data and we
+        // need it to easily hook into the expression templates.
+        logical_du.set_data_ref(const_cast<double*>(  // NOLINT
+                                    gsl::at(logical_partial_derivatives_of_u,
+                                            logical_deriv_index)) +  // NOLINT
+                                    component_index * num_grid_points,
+                                num_grid_points);
+        lhs +=
+            (*(inverse_jacobian.begin() +
+               gsl::at(gsl::at(indices, logical_deriv_index), deriv_index))) *
+            logical_du;
+      }
+      // clang-tidy: no pointer arithmetic
+      pdu += num_grid_points;  // NOLINT
+    }
+  }
+}
 }  // namespace partial_derivatives_detail
 
+template <typename DerivativeTags, typename VariableTags, size_t Dim>
+void logical_partial_derivatives(
+    const gsl::not_null<std::array<Variables<DerivativeTags>, Dim>*>
+        logical_partial_derivatives_of_u,
+    const Variables<VariableTags>& u, const Mesh<Dim>& mesh) noexcept {
+  if (UNLIKELY((*logical_partial_derivatives_of_u)[0].number_of_grid_points() !=
+               u.number_of_grid_points())) {
+    for (auto& deriv : *logical_partial_derivatives_of_u) {
+      deriv.initialize(u.number_of_grid_points());
+    }
+  }
+  std::array<double*, Dim> deriv_pointers{};
+  for (size_t i = 0; i < Dim; ++i) {
+    gsl::at(deriv_pointers, i) =
+        gsl::at(*logical_partial_derivatives_of_u, i).data();
+  }
+  if (Dim == 1) {
+    Variables<DerivativeTags>* temp = nullptr;
+    partial_derivatives_detail::LogicalImpl<Dim, VariableTags, DerivativeTags>::
+        apply(make_not_null(&deriv_pointers), temp, u, mesh);
+  }
+  Variables<DerivativeTags> temp(u.number_of_grid_points());
```

Comment:
Lines 129–132 look a lot like 125–127. Are 129–132 intended to be in an `else` block? If this is as intended, please add clarifying comment.